src/sinks/datadog/logs/sink.rs

```diff
@@ -0,0 +1,394 @@
+use crate::sinks::util::sink::{Response, ServiceLogic, StdServiceLogic};
+use crate::sinks::util::Compression;
+use async_trait::async_trait;
+use flate2::write::GzEncoder;
+use futures::future::FutureExt;
+use futures::stream::BoxStream;
+use futures::{StreamExt, TryFutureExt};
+use futures_util::stream::FuturesUnordered;
+use metrics::gauge;
+use snafu::Snafu;
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::hash::BuildHasherDefault;
+use std::io::Write;
+use std::num::NonZeroUsize;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::mpsc::{channel, Receiver};
+use tokio::sync::oneshot;
+use tokio::{pin, select};
+use tower::{Service, ServiceExt};
+use tracing_futures::Instrument;
+use twox_hash::XxHash64;
+use vector_core::buffers::Acker;
+use vector_core::config::{log_schema, LogSchema};
+use vector_core::event::{Event, EventFinalizers, Finalizable, Value};
+use vector_core::partition::Partitioner;
+use vector_core::sink::StreamSink;
+use vector_core::stream::batcher::Batcher;
+
+const MAX_PAYLOAD_ARRAY: usize = 1_000;
+const MAX_PAYLOAD_BYTES: usize = 5_000_000;
+
+#[derive(Default)]
+struct EventPartitioner {}
+
+impl Partitioner for EventPartitioner {
+    type Item = Event;
+    type Key = Option<Arc<str>>;
+
+    fn partition(&self, item: &Self::Item) -> Self::Key {
+        item.metadata().datadog_api_key().clone()
+    }
+}
+
+#[derive(Debug)]
+pub struct LogSinkBuilder<S> {
+    encoding: EncodingConfigWithDefault<Encoding>,
+    service: S,
+    context: SinkContext,
+    timeout: Option<Duration>,
+    compression: Option<Compression>,
+    default_api_key: Option<Arc<str>>,
+    log_schema: Option<&'static LogSchema>,
+}
+
+impl<S> LogSinkBuilder<S> {
+    pub fn new(service: S, context: SinkContext) -> Self {
+        Self {
+            encoding: Default::default(),
+            service,
+            context,
+            default_api_key: None,
+            timeout: None,
+            compression: None,
+            log_schema: None,
+        }
+    }
+
+    pub const fn log_schema(mut self, log_schema: &'static LogSchema) -> Self {
+        self.log_schema = Some(log_schema);
+        self
+    }
+
+    #[allow(clippy::missing_const_for_fn)] // const cannot run destructor
+    pub fn encoding(mut self, encoding: EncodingConfigWithDefault<Encoding>) -> Self {
+        self.encoding = encoding;
+        self
+    }
+
+    pub const fn compression(mut self, compression: Compression) -> Self {
+        self.compression = Some(compression);
+        self
+    }
+
+    pub const fn batch_timeout(mut self, duration: Option<Duration>) -> Self {
+        self.timeout = duration;
+        self
+    }
+
+    pub fn default_api_key(mut self, api_key: Arc<str>) -> Self {
+        self.default_api_key = Some(api_key);
+        self
+    }
+
+    pub fn build(self) -> LogSink<S> {
+        LogSink {
+            default_api_key: self.default_api_key.unwrap(),
```

Comment:
Related to my comment above, the builder allows for the possibility of the API key to not be specified, but it would blow up here with a panic.  These two things don't seem to line up.

Reply:
You're quite right. Resolved in f37fbcbfc